rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp

```diff
@@ -62,6 +64,21 @@ SequentialCompressionWriter::~SequentialCompressionWriter()
 
 void SequentialCompressionWriter::compression_thread_fn()
 {
+  if (compression_options_.thread_nice_value) {
+    uint8_t nice_value = *compression_options_.thread_nice_value;
+
+    // according to posix this sets the nice value for the whole process
+    // but in reality this sets the thread priority on linux.
+    // There does not seem to be any other way to set the thread priority on
+    // a non-realtime kernel :-(
+    // https://stackoverflow.com/questions/10876342/equivalent-of-setthreadpriority-on-linux-pthreads
+    if (0 != setpriority(PRIO_PROCESS, 0, nice_value)) {
```

Comment:
Does this work on Windows?

Reply:
Also, I'm not convinced that this is the correct place to do this (even though it is hidden behind an option).  If you compose this node with other nodes, then you'll be setting the niceness of the whole process, not just this node.  That may be your intention, but in that case why not just use the system tools (like `nice`) to do this?  Can you describe your use case in more detail?

Reply:
good point, I'll add an ifdef version....

Reply:
> Also, I'm not convinced that this is the correct place to do this (even though it is hidden behind an option). If you compose this node with other nodes, then you'll be setting the niceness of the whole process, not just this node. That may be your intention, but in that case why not just use the system tools (like `nice`) to do this? Can you describe your use case in more detail?

Its only setting the nice value of the calling thread, not the whole process, see
man 2 setpriority 
section BUGS

So the use case is, that while we dump our data into the logdata, we want the main thread to continue receiving data with normal scheduling priority. Also we don't want other processes, to get less scheduler time due to the writing of the log file.

Reply:
I added an ifdef for windows, and an implementation that should work, but which is commented, as I can not test it due to a lack of a windows...

For the linux implementation I switched to the nice function, which by documentation only modifies the current thread:
`DESCRIPTION
       nice() adds inc to the nice value for the calling thread.  (A higher nice value means a lower priority.)
`
, without using dubious BUGS in the posix implementation.

Reply:
@jmachowinski It is a bit of a hacky and non-deterministic way to manage resources.  
The `nice` and `setpriority` actually doing exactly the same. The reason why you see in Linux documentation that nice changing niceness for the thread rather than a priority is not in a different implementation of the `nice` for Linux. 
It is just because of another issue that Linux threads implementation doesn't follow the POSIX.1 standard.
See details at the https://stackoverflow.com/a/7684821

The more appropriate and deterministic way to manage resources would be:
1. Run rosbag2 recorder in a separate process
2. Run rosbag2 recorder on separate isolated CPU cores to avoid interference with other processes.

Reply:
I should elaborate on the context of this commit.
Our use case is an industrial system, that has a node, that records messages from various topics into a **memory limited** ringbuffer. (https://github.com/cellumation/snapshotter) The reception of the messages is done in the main thread, and this should have a normal scheduler priority, so that we don't drop messages. 
For debugging of problems, the customers can trigger a so called snapshot, that dumps the ringbuffer to the disk.
This is done, while the machine is running, therefore we do not want the compression and writing thread to have a high priority, to not interfere with the rest of the system.

Note, that we do not use rosbag2 directly, as the memory overhead of saving the topic name with each message is not acceptable for our use case.

@MichaelOrlov 
> Run rosbag2 recorder in a separate process

In this case we would eventually drop messages as the reception thread would also have a high nice level.

>  Run rosbag2 recorder on separate isolated CPU cores to avoid interference with other processes.

We don't have enough CPUs for that.

Anyhow, I just found out, that the nice level of a new thread is inherited from its parent, and that therefore I could drop it in my code before calling open. Therefore I would be fine with dropping this PR if you don't think this in general a useful feature.

Reply:
@jmachowinski Thanks for the explanation and details.  
Now it makes sense, I see the point. Let's work together and try to come to some solution for that.

Reply:
@jmachowinski BTW as regards
> Note, that we do not use rosbag2 directly, as the memory overhead of saving the topic name with each message is not acceptable for our use case.

With the new MCAP storage format, we are not storing the topic name with each message.
See detail in the
https://github.com/ros2/rosbag2/blob/dcd239ceea842752c344f0048572deb5f502feb8/rosbag2_storage_mcap/src/mcap_storage.cpp#L732-L742

Reply:
> @jmachowinski BTW as regards
> 
> With the new MCAP storage format, we are not storing the topic name with each message. See detail in the
> 

I meant the MessageCacheInterface, which is bound to the rosbag2_storage::SerializedBagMessage.

I could not see a way of implementing my own buffer, without using rosbag2_storage::SerializedBagMessage.
Note, I was talking about the **in memory** overhead, which is significant, if you have a lot of small messages and per message the topic name is saved.

Here you can find a comparison of memory usage for the ros1 version, were we ran into the same issue but with the ShapeShifterMsg : 
https://github.com/cellumation/snapshotter/tree/ros1

Reply:
@jmachowinski Thanks for the input.  
I understand that in theory if we would store some index to the `topic_name` instead of strings it would save us potentially some memory.  
However, performance measurements results from the https://github.com/cellumation/snapshotter/tree/ros1#performance-evaluation-and-comparision a bit confusing.
It looks like `rosbag2_snapshot` implementation consumes 200 MBytes less than `snapsotter`

<h4 tabindex="-1" dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 16px; font-weight: var(--base-text-weight-semibold, 600); line-height: 1.25; color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Scenario 2</h4><p dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Run a gazebo simulation with 5 depth cameras and several small topics. Limit memory usage to 10GB.</p>

  | rosbag_snapshot | snapshotter
-- | -- | --
cpu usage avg % | 56% | 52%
Ram usage | 9.8GB | 10.0GB
Resulting Bag Size | 8.4GB | 9.8GB
Resulting Bag Duration | 9.6s | 10.7s

<h4 tabindex="-1" dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 16px; font-weight: var(--base-text-weight-semibold, 600); line-height: 1.25; color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><a id="user-content-scenario-3" class="anchor" aria-hidden="true" tabindex="-1" href="https://github.com/cellumation/snapshotter/tree/ros1#scenario-3" style="box-sizing: border-box; background-color: transparent; color: var(--fgColor-accent, var(--color-accent-fg)); text-decoration: none; float: left; padding-right: 4px; margin-left: -20px; line-height: 1; position: absolute;"><svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path d="m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z"></path></svg></a></h4>

<h4 tabindex="-1" dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 16px; font-weight: var(--base-text-weight-semibold, 600); line-height: 1.25; color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Scenario 3</h4><p dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Run a gazebo simulation with only small topics. Limit memory usage to 10GB.</p>

  | rosbag_snapshot | snapshotter
-- | -- | --
cpu usage avg % | 10% | 8%
Ram usage | 9.8GB | 10.0GB
Resulting Bag Size | 6.8GB | 9.8GB
Resulting Bag Duration | 619s | 954s

<h4 tabindex="-1" dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 16px; font-weight: var(--base-text-weight-semibold, 600); line-height: 1.25; color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><a id="user-content-results" class="anchor" aria-hidden="true" tabindex="-1" href="https://github.com/cellumation/snapshotter/tree/ros1#results" style="box-sizing: border-box; background-color: transparent; color: var(--fgColor-accent, var(--color-accent-fg)); text-decoration: none; float: left; padding-right: 4px; margin-left: -20px; line-height: 1; position: absolute;"><svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path d="m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z"></path></svg></a></h4>

Reply:
> It looks like rosbag2_snapshot implementation consumes 200 MBytes less than snapsotter

@arneboe can you comment on this, you did the tests back in the day...

My guess would be, it is a bug in rosback_snapshot and both programs were given a parameter to run  with a memory limit of 10 GB.

Reply:
@MichaelOrlov This is a test between https://github.com/ros/rosbag_snapshot and our snapshotter, not rosbag2_snapshot. Note, the results should be still comparable, to give an idea of the overhead created by the topic name.